src/gpuhunt/providers/cudo.py

```diff
@@ -22,94 +29,309 @@ class CudoProvider(AbstractProvider):
+    # Add disk
+    unbalanced_specs = [
+        {
+            "cpu": spec["cpu"],
+            "memory": spec["memory"],
+            "gpu": 0,
+            "disk_size": min_none(machine_type["maxStorageGibFree"], disk_size),
+        }
+        for spec in unbalanced_specs
+    ]
+    return unbalanced_specs
+
+
+def get_cpu_range(min_cpu, max_cpu, max_cpu_free):
+    cpu_range = range(
+        min_cpu if min_cpu is not None else MIN_CPU,
+        min(max_cpu if max_cpu is not None else max_cpu_free, max_cpu_free) + 1,
+    )
+    return cpu_range
 
-class VMTypeFetchError(Exception):
-    def __init__(self, message, vcpu, memory_gib, gpu):
-        super().__init__(message)
-        self.vcpu = vcpu
-        self.memory_gib = memory_gib
-        self.gpu = gpu
 
-    def __str__(self):
-        return f"{super().__str__()} - [vCPU: {self.vcpu}, Memory: {self.memory_gib} GiB, GPU: {self.gpu}]"
+def get_gpu_range(min_gpu_count, max_gpu_count, max_gpu_free):
+    gpu_range = range(
+        min_gpu_count if min_gpu_count is not None else 1,
+        min(max_gpu_count if max_gpu_count is not None else max_gpu_free, max_gpu_free) + 1,
+    )
+    return gpu_range
+
+
+def get_memory_range(min_memory, max_memory, max_memory_gib_free):
+    memory_range = range(
+        int(min_memory) if min_memory is not None else MIN_MEMORY,
+        min(
+            int(max_memory) if max_memory is not None else max_memory_gib_free, max_memory_gib_free
+        )
+        + 1,
+    )
+    return memory_range
+
+
+def get_balanced_memory(gpu_count, gpu_memory, max_memory):
+    return min_none(
+        round_up(RAM_PER_VRAM * gpu_memory * gpu_count, RAM_DIV), round_down(max_memory, RAM_DIV)
+    )
+
+
+def get_balanced_cpu(memory, max_cpu):
+    return min_none(
+        round_up(ceil(memory / RAM_PER_CORE), CPU_DIV),
+        round_down(max_cpu, CPU_DIV),  # can be None
+    )
+
+
+def get_balanced_disk_size(available_disk, memory, total_gpu_memory, max_disk_size, min_disk_size):
+    return max_none(
+        min_none(
+            available_disk,
+            max(memory, total_gpu_memory),
+            max_disk_size,
+        ),
+        min_disk_size,
+    )
 
 
 def gpu_name(name: str) -> Optional[str]:
     if not name:
         return None
     result = GPU_MAP.get(name)
-    if result is None:
-        raise Exception("There is no '%s' in GPU_MAP", name)
     return result
 
 
 def get_memory(gpu_name: str) -> Optional[int]:
+    if not gpu_name:
+        return None
     for gpu in KNOWN_GPUS:
         if gpu.name.lower() == gpu_name.lower():
             return gpu.memory
-    raise Exception("There is no '%s' in KNOWN_GPUS", gpu_name)
+
+
+def round_up(value: Optional[Union[int, float]], step: int) -> Optional[int]:
+    if value is None:
+        return None
+    return round_down(value + step - 1, step)
+
+
+def round_down(value: Optional[Union[int, float]], step: int) -> Optional[int]:
```

Comment:
VSCode suggests that round_up and round_down may return `Optional[Union[float, int]]`

Reply:
Yes. I think  returning int(value // step * step) in round_down() will solve. What do yo suggest? These functions are used in tensordock too. Shall I change there too?

Reply:
Yes! Thanks! 😀